File: baekjoon/N1014.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class N1014 {
	static int n, m, memo[][][];
	static boolean map[][];

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int c = Integer.parseInt(br.readLine());
		for (int i = 0; i < c; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			n = Integer.parseInt(st.nextToken());
			m = Integer.parseInt(st.nextToken());
			map = new boolean[n][m + 1];
			memo = new int[n][m + 1][1 << (m + 2)];
			for (int j = 0; j < n; j++) {
				char in[] = br.readLine().toCharArray();
				for (int k = 1; k <= m; k++) {
					map[j][k] = in[k - 1] == '.';
					Arrays.fill(memo[j][k], -1);
				}
			}
			System.out.println(dp(0, 1, 0));
		}
	}

	static int dp(int a, int b, int c) {
		if (b > m) {
			if (a == n - 1) {
				return 0;
			}
			return dp(a + 1, 1, c);
		}
		if (memo[a][b][c] < 0) {
			memo[a][b][c] = dp(a, b + 1, c - (c & (1 << b)));
			if (map[a][b] && (c & (1 << b)) == 0) {
				memo[a][b][c] = Math.max(memo[a][b][c], dp(a, b + 2, c | (1 << (b - 1)) | (1 << (b + 1))) + 1);
			}
		}
		return memo[a][b][c];
	}
}

File: baekjoon/N1014.cpp
#include <stdio.h>
int m, n;
bool in[11][10];
int memo[11][10][1 << 12];

int dp(int x, int y, int bm)
{
	if (x > m)
	{
		if (y == n - 1)
			return 0;
		return dp(1, y + 1, bm);
	}
	if (memo[x][y][bm] > -1)
		return memo[x][y][bm];
	int max = dp(x + 1, y, bm & ~(1 << x));
	if (!(bm & (1 << x)) && in[x][y])
		max = max > dp(x + 1, y, bm | 1 << (x - 1) | 1 << (x + 1)) + 1 ? max : dp(x + 1, y, bm | 1 << (x - 1) | 1 << (x + 1)) + 1;
	return memo[x][y][bm] = max;
}

int main()
{
	int c;
	char s[10];
	scanf("%d", &c);
	while (c--)
	{
		scanf("%d %d", &n, &m);
		for (int i = 0; i < n; i++)
		{
			scanf("%s", s);
			for (int j = 1; j <= m; j++)
			{
				in[j][i] = s[j - 1] == '.';
				for (int k = 0; k < (1 << 12); k++)
					memo[j][i][k] = -1;
			}
		}
		printf("%d\n", dp(1, 0, 0));
	}
	return 0;
}
